src/main/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java

```diff
@@ -509,28 +522,98 @@ public static synchronized void initializeAnyploidPLIndexToAlleleIndices(final i
+         * Creates the ordering for all PL indices less than or equal to PLIndex, and stores them for future use.
+         *
+         * The implementation is described in The Variant Call Format Specification VCF 4.3, Section 1.6.2 Genotype fields
+         * The likelihoods are ordered for ploidy P and N alternate alleles as follows:
+         * for aP = 0...N
+         *  for aP-1 = 0...aP
+         *      ...
+         *      for a1 = 0...a2
+         *          a1,a2..aP
+         *
+         * However, instead of having a set number of alternate alleles, we allow the outer loop to continue indefinitely, and break out
+         * once the correct PLIndex has been reached.  The state of the looping is stored, so that calculating higher indices can start
+         * from where we left off, instead of going back the the beginning.
+         *
+         * To do this, we do not keep track of the indices of each of the P loops explicitly.  In the loop pseudocode, we note that writing a new allele list
+         * corresponds to incrementing one loop, leaving the indices in the loops above it unchanged, and resetting the indices in the loops below it to zero.
+         * After incrementing the index of a particular loop, we next increment the index of the loop just below it, unless we have just incremented the lowest loop,
+         * in which case we next incremented whichever is the lowest loop that can still be incremented.  Thus, we note that we can write the list of alleles for the
+         * PLIndex pl_i based on the PLIndex pl_i - 1 if we know which single allele must be incremented. If we know we incremented allele a_i to
+         * create the list of alleles for pl_i - 1 from the list of alleles for pl_i - 2, then, if the index a_i is not 0 (ie. the leftmost allele),
+         * we simply increment allele a_i - 1 to create the list of alleles for pl_i, and leave the other alleles unchanged (by construction, only alleles to
+         * the right of allele a_i - 1 will be nonzero).  If allele a_i - 1 is less than allele a_i, then we know that it can further be incremented before
+         * allele a_i, so we add the index a_i to a deque D_a (by construction, D_a will always be sorted in descending order).  If the index a_i is 0, then we poll from the end of D_a to find the next allele to increment
+         * (or set the next allele to ploidy - 1, ie the rightmost allele,  if D_a is empty).  In this case, all alleles to the left of the incremented allele
+         * are reset to 0, and all alleles to the right are left unchanged.
+         *
+         * The algorithm is initialized in thr constructor for PLIndex 0 with a list of P 0's and by setting the next allele to increment as ploidy - 1 (ie, the rightmost allele).
+         * @param PLIndex
+         * @return
+         */
+
+        private List<Integer> getAlleles(final int PLIndex) {
+
+            while(PLIndex >= allelesList.size()) {
```

Comment:
This looks good to me but I don't have high confidence that I'm able to find errors in it by inspection.  There are a lot of possible off-by-one error sites.

Reply:
Agreed on the possibility for off-by-one errors.  I think the large "derived in reverse" tests in `GenotypeLikelihoodsUnitTests::testGetAllelesData` are a pretty good guard against this, and I added a bunch more randomly generated similar tests to be safe.  It seems quite unlikely that none of them would fail if there were an off-by-one error floating around here.

Reply:
That was my thought too.  If it works empirically for a large set of things then it should continue to work.  It doesn't have a lot of obvious edge cases that don't also get hit on the way to a larger number.